File: sdks/livekit-js/src/export.ts
import { createHash, timingSafeEqual } from "node:crypto";

import {
  ProxyTracerProvider,
  trace,
  type Attributes,
  type TracerProvider,
} from "@opentelemetry/api";
import { OTLPTraceExporter } from "@opentelemetry/exporter-trace-otlp-proto";
import {
  defaultResource,
  resourceFromAttributes,
} from "@opentelemetry/resources";
import {
  BatchSpanProcessor,
  NodeTracerProvider,
  type SpanExporter,
  type SpanProcessor,
} from "@opentelemetry/sdk-trace-node";
import { ATTR_SERVICE_NAME } from "@opentelemetry/semantic-conventions";
import { telemetry, type JobContext } from "@livekit/agents";

/**
 * The one Egma exporter in this process, shared by both verbs.
 *
 * `monitor` and `simulation` send the same thing to the same door: this
 * worker's OpenTelemetry spans, over OTLP, with the project API key on them.
 * They differ in one fact — a simulation stamps the room it runs in on every
 * span, so Egma can file the agent's POV under the simulation that room
 * belongs to.
 *
 * **Why the room name rides two ways at once.** A resource is fixed when a
 * tracer provider is built, and this SDK does not always build one: a worker
 * that already has OpenTelemetry set up hands it a provider that exists, and
 * that provider must keep working. So the room name goes on twice — as a
 * resource attribute where this SDK builds the provider, and as a span
 * attribute on every span through LiveKit's own `setTracerProvider` metadata
 * seam, which works on any provider whoever built it. Egma's door reads the
 * resource first and falls back to the spans when every span in a resource
 * agrees.
 *
 * **One job per process.** The room this process exports under is decided by
 * the first job that asks, because a resource cannot be rewritten and the
 * metadata processor goes on once. LiveKit runs one job per process, and a
 * second job asking for different settings is refused here rather than filed
 * under the first job's name.
 */

const TRACE_PATH = "/v1/traces";
const PROJECT_KEY_PATTERN = /^egma_sk_[A-Za-z0-9_-]{43}$/u;
const UNCONFIGURED_TRACER_PROVIDER = new ProxyTracerProvider().getDelegate();

/**
 * What Egma files a simulation's agent POV under: the room's name.
 *
 * Egma's own attribute, in Egma's own namespace, so it can never collide with
 * a semantic convention or a framework's own key. A resource without it is
 * production traffic and takes the path it always took.
 */
export const PROVIDER_REFERENCE = "egma.provider_reference";

/**
 * How long a simulation's spans may sit in the buffer: one second.
 *
 * Short because somebody is waiting. A simulation is graded the moment the
 * agent's POV is complete, so the tail of the conversation has to land within
 * a second or two of the persona leaving. Production is not waited on the same
 * way and keeps the library's own default.
 */
export const SIMULATION_BATCH_MILLIS = 1_000;

type CloudSpanProcessorOptions = {
  readonly url: string;
  readonly headers: Record<string, string>;
  readonly exporter?: SpanExporter;
};

type CreateCloudSpanProcessor = (
  options: CloudSpanProcessorOptions,
) => SpanProcessor;

type MutableFanoutSpanProcessor = SpanProcessor & {
  add(processor: SpanProcessor): void;
};

type SharedLiveKitTelemetry = {
  readonly FanoutSpanProcessor: new () => MutableFanoutSpanProcessor;
  readonly tracer: typeof telemetry.tracer & {
    getProvider(): TracerProvider;
  };
  readonly setTracerProvider: (
    provider: TracerProvider,
    options?: {
      readonly metadata?: Attributes;
      readonly registerSpanProcessor?: (processor: SpanProcessor) => void;
      readonly createCloudSpanProcessor?: CreateCloudSpanProcessor;
    },
  ) => void;
};

export type ExportState = {
  readonly verb: string;
  readonly endpoint: string;
  readonly apiKeyDigest: Buffer;
  readonly roomName: string;
  readonly providerReference: string;
  readonly provider: TracerProvider;
  readonly processor: BatchSpanProcessor;
  readonly registerSpanProcessor: (processor: SpanProcessor) => void;
};

let state: ExportState | undefined;
let contextsWithFlush = new WeakSet<object>();

/**
 * The mutable seam for a tracer provider that another integration already
 * installed. OpenTelemetry JS 2.x cannot add processors to a provider after
 * construction, so the provider must have been built around a fan-out
 * processor and this callback must add to that exact fan-out.
 */
export type ExistingTelemetry = {
  readonly provider: TracerProvider;
  readonly registerSpanProcessor: (processor: SpanProcessor) => void;
  readonly createCloudSpanProcessor?: CreateCloudSpanProcessor;
};

export type ExportOptions = {
  /** Egma API origin. Defaults to `EGMA_URL`. */
  readonly endpoint?: string;
  /** Egma project API key. Defaults to `EGMA_API_KEY`. */
  readonly apiKey?: string;
  readonly existingTelemetry?: ExistingTelemetry;
};

/**
 * Export this worker's spans to Egma, once per process.
 *
 * `providerReference` is the room a simulation runs in, or `""` for
 * production. Returns the processor, so a caller with a second moment worth
 * flushing at — a session closing — can ask for one.
 */
export function installExport(
  ctx: JobContext,
  options: ExportOptions,
  verb: string,
  providerReference: string,
): BatchSpanProcessor {
  const roomName = jobRoomName(ctx, verb);
  const sharedTelemetry = sharedLiveKitTelemetry();
  if (sharedTelemetry === undefined) {
    throw new Error(
      `${verb} requires a supported @livekit/agents version (>=1.5.5 <2) that exposes the shared telemetry seam Egma needs.`,
    );
  }

  const addShutdownCallback = contextShutdownCallback(ctx, verb);
  const endpoint = traceEndpoint(setting(options.endpoint, "EGMA_URL", verb), verb);
  const apiKey = projectKey(setting(options.apiKey, "EGMA_API_KEY", verb), verb);
  const apiKeyDigest = createHash("sha256").update(apiKey).digest();

  if (state === undefined) {
    state = configureExport(
      endpoint,
      apiKey,
      apiKeyDigest,
      roomName,
      providerReference,
      jobAgentName(ctx),
      verb,
      sharedTelemetry,
      options.existingTelemetry,
    );
  } else if (
    state.endpoint !== endpoint ||
    !timingSafeEqual(state.apiKeyDigest, apiKeyDigest)
  ) {
    throw new Error(
      `${verb} is already configured with different settings in this process. Restart the worker after changing EGMA_URL or EGMA_API_KEY.`,
    );
  } else if (state.roomName !== roomName || state.verb !== verb) {
    // Neither room is quoted. A provider's resource is fixed when it is
    // built, so what a reader has to act on is the arrangement rather than
    // which two rooms collided, and an error that names a customer's rooms
    // is one more place they can travel to.
    throw new Error(
      `${verb} is already configured for a different LiveKit job in this process. Restart the worker so each LiveKit job keeps its own room and its own trace metadata: a provider's resource is fixed when it is built, so LiveKit must run one job per process.`,
    );
  } else if (
    options.existingTelemetry !== undefined &&
    state.provider !== options.existingTelemetry.provider
  ) {
    throw new Error(
      `${verb} is already configured with a different OpenTelemetry tracer provider in this process. Restart the worker after changing tracing setup.`,
    );
  }

  registerShutdownFlush(ctx, addShutdownCallback, state.processor);
  return state.processor;
}

/** Send whatever is buffered now, and never let the failure stop the job. */
export async function flushNow(
  processor: BatchSpanProcessor,
  why: string,
): Promise<void> {
  try {
    await processor.forceFlush();
  } catch {
    console.warn(
      `Egma: could not flush every buffered span at ${why}.`,
    );
  }
}

function sharedLiveKitTelemetry(): SharedLiveKitTelemetry | undefined {
  const candidate = telemetry as unknown as {
    readonly FanoutSpanProcessor?: unknown;
    readonly setTracerProvider?: unknown;
    readonly tracer?: { readonly getProvider?: unknown };
  };
  if (
    typeof candidate.FanoutSpanProcessor !== "function" ||
    typeof candidate.setTracerProvider !== "function" ||
    typeof candidate.tracer?.getProvider !== "function"
  ) {
    return undefined;
  }
  return telemetry as unknown as SharedLiveKitTelemetry;
}

function jobRoomName(ctx: JobContext, verb: string): string {
  const roomName = ctx?.job?.room?.name;
  if (typeof roomName !== "string") {
    throw new Error(
      `${verb} needs the LiveKit JobContext this job was given.`,
    );
  }
  return roomName;
}

function contextShutdownCallback(
  ctx: JobContext,
  verb: string,
): (callback: () => Promise<void>) => void {
  const callback = ctx?.addShutdownCallback;
  if (typeof callback !== "function") {
    throw new Error(
      `${verb} needs the LiveKit JobContext this job was given.`,
    );
  }
  return callback.bind(ctx);
}

function jobAgentName(ctx: JobContext): string {
  const agentName = ctx?.job?.agentName;
  return typeof agentName === "string" ? agentName : "";
}

function setting(
  explicit: string | undefined,
  environmentName: string,
  verb: string,
): string {
  const value = explicit ?? process.env[environmentName];
  if (typeof value !== "string" || value.trim() === "") {
    throw new Error(
      `${verb} needs ${environmentName}. Set it or pass the matching option.`,
    );
  }
  return value.trim();
}

export function projectKey(value: string, verb = "Egma"): string {
  if (!PROJECT_KEY_PATTERN.test(value)) {
    throw new Error(`${verb} received an invalid EGMA_API_KEY.`);
  }
  return value;
}

export function traceEndpoint(value: string, verb = "Egma"): string {
  let parsed: URL;
  try {
    parsed = new URL(value);
  } catch {
    throw invalidEndpoint(verb);
  }

  if (
    (parsed.protocol !== "http:" && parsed.protocol !== "https:") ||
    parsed.hostname === "" ||
    parsed.username !== "" ||
    parsed.password !== "" ||
    parsed.search !== "" ||
    parsed.hash !== "" ||
    /\s/u.test(value)
  ) {
    throw invalidEndpoint(verb);
  }

  const basePath = parsed.pathname.replace(/\/+$/u, "");
  parsed.pathname = basePath.endsWith(TRACE_PATH)
    ? basePath
    : `${basePath}${TRACE_PATH}`;
  return parsed.toString();
}

function invalidEndpoint(verb: string): Error {
  return new Error(
    `${verb} needs EGMA_URL to be a valid HTTP or HTTPS API URL.`,
  );
}

function configureExport(
  endpoint: string,
  apiKey: string,
  apiKeyDigest: Buffer,
  roomName: string,
  providerReference: string,
  agentName: string,
  verb: string,
  sharedTelemetry: SharedLiveKitTelemetry,
  suppliedTelemetry: ExistingTelemetry | undefined,
): ExportState {
  const existingTelemetry = compatibleExistingTelemetry(
    sharedTelemetry.tracer.getProvider(),
    trace.getTracerProvider(),
    suppliedTelemetry,
    verb,
  );

  let processor: BatchSpanProcessor | undefined;
  try {
    const exporter = new OTLPTraceExporter({
      url: endpoint,
      headers: { Authorization: `Bearer ${apiKey}` },
    });
    processor =
      providerReference === ""
        ? new BatchSpanProcessor(exporter)
        : new BatchSpanProcessor(exporter, {
            scheduledDelayMillis: SIMULATION_BATCH_MILLIS,
          });
    const ownedFanout = new sharedTelemetry.FanoutSpanProcessor();
    let provider: TracerProvider;
    let registerSpanProcessor: (added: SpanProcessor) => void;
    let createCloudSpanProcessor =
      existingTelemetry?.createCloudSpanProcessor;

    if (existingTelemetry === undefined) {
      const ownedProvider = new NodeTracerProvider({
        resource: defaultResource().merge(
          resourceFromAttributes({
            [ATTR_SERVICE_NAME]: "livekit-agents",
            // A resource this SDK builds can carry the reference, and that
            // is the copy Egma's door reads first.
            ...(providerReference === ""
              ? {}
              : { [PROVIDER_REFERENCE]: providerReference }),
          }),
        ),
        spanProcessors: [processor, ownedFanout],
      });
      ownedProvider.register();
      provider = ownedProvider;
      registerSpanProcessor = (added) => ownedFanout.add(added);
      createCloudSpanProcessor = ({ url, headers, exporter: supplied }) =>
        new BatchSpanProcessor(
          supplied ?? new OTLPTraceExporter({ url, headers }),
        );
    } else {
      provider = existingTelemetry.provider;
      registerSpanProcessor = existingTelemetry.registerSpanProcessor;
      registerSpanProcessor(processor);
    }
    sharedTelemetry.setTracerProvider(provider, {
      metadata: {
        "session.id": roomName,
        // The copy that works whoever built the provider: LiveKit stamps
        // this on every span it starts.
        ...(providerReference === ""
          ? {}
          : { [PROVIDER_REFERENCE]: providerReference }),
        ...(agentName === ""
          ? {}
          : { [telemetry.traceTypes.ATTR_AGENT_NAME]: agentName }),
      },
      registerSpanProcessor,
      ...(createCloudSpanProcessor === undefined
        ? {}
        : {
            createCloudSpanProcessor,
          }),
    });

    return {
      verb,
      endpoint,
      apiKeyDigest,
      roomName,
      providerReference,
      provider,
      processor,
      registerSpanProcessor,
    };
  } catch {
    void processor?.shutdown().catch(() => undefined);
    throw new Error(
      `${verb} could not configure the Egma exporter. Check EGMA_URL, EGMA_API_KEY, and the worker's OpenTelemetry setup.`,
    );
  }
}

function compatibleExistingTelemetry(
  liveKitProvider: TracerProvider,
  globalProvider: TracerProvider,
  supplied: ExistingTelemetry | undefined,
  verb: string,
): ExistingTelemetry | undefined {
  const configured = [liveKitProvider, globalProvider]
    .map(configuredProvider)
    .filter((provider): provider is TracerProvider => provider !== undefined);
  const distinct = [...new Set(configured)];

  if (distinct.length === 0) {
    if (supplied === undefined) return undefined;
    throw incompatibleProvider(verb);
  }
  if (
    distinct.length > 1 ||
    supplied === undefined ||
    distinct[0] !== supplied.provider
  ) {
    throw incompatibleProvider(verb);
  }
  return supplied;
}

function incompatibleProvider(verb: string): Error {
  return new Error(
    `${verb} found an existing OpenTelemetry tracer provider that it cannot safely extend. Pass existingTelemetry with that provider and its span-processor registrar, or call ${verb} before custom tracing setup.`,
  );
}

function configuredProvider(
  provider: TracerProvider,
): TracerProvider | undefined {
  if (!(provider instanceof ProxyTracerProvider)) return provider;
  const delegate = provider.getDelegate();
  return delegate === UNCONFIGURED_TRACER_PROVIDER ? undefined : delegate;
}

function registerShutdownFlush(
  ctx: JobContext,
  addShutdownCallback: (callback: () => Promise<void>) => void,
  processor: BatchSpanProcessor,
): void {
  if (contextsWithFlush.has(ctx)) return;

  addShutdownCallback(async () => {
    await flushNow(processor, "job shutdown");
  });
  contextsWithFlush.add(ctx);
}

/** @internal Test-only access; not exported from the package root. */
export function exportStateForTests(): ExportState | undefined {
  return state;
}

/** @internal Test-only reset; not exported from the package root. */
export function resetExportForTests(): void {
  state = undefined;
  contextsWithFlush = new WeakSet<object>();
}

File: sdks/python/src/egma/export.py
"""LiveKit's OTLP exporter for ``monitor`` and ``simulation``.

Use the project API key and export the simulation room as its provider
reference. Set that reference on resources when creating a provider and
on spans through LiveKit metadata when extending an existing provider.
This preserves the customer's tracing setup, whose resource is immutable.

Ingestion prefers the resource value, falling back to matching span
values. One job per process is supported: the first job fixes exporter
settings, and another job requesting different settings is refused.
The framework-free parts live in ``egma.otlp``.
"""

from __future__ import annotations

import hashlib
import logging
import threading
from dataclasses import dataclass

from livekit.agents.telemetry import set_tracer_provider
from opentelemetry import trace
from opentelemetry.sdk.resources import SERVICE_NAME, Resource
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor, SpanExporter

from . import otlp
from .otlp import PROVIDER_REFERENCE, SIMULATION_BATCH_MILLIS, flush_on

__all__ = [
    "PROVIDER_REFERENCE",
    "SIMULATION_BATCH_MILLIS",
    "flush_on",
    "install",
]

logger = logging.getLogger("egma")

_FLUSH_MARKER = "_egma_export_flush_registered"

_ROOT_SPAN = "agent_session"
"""LiveKit's session span: the end of a simulation's agent POV."""


@dataclass(frozen=True)
class _Export:
    """The one Egma exporter configured in this process, and its terms."""

    verb: str
    endpoint: str
    api_key_digest: bytes
    provider_reference: str
    provider: TracerProvider
    processor: BatchSpanProcessor


_state: _Export | None = None
_state_lock = threading.Lock()


class _SimulationEvidenceExporter(otlp.RootLastExporter):
    """Keep a simulation completion root behind all earlier evidence."""

    def __init__(self, delegate: SpanExporter) -> None:
        super().__init__(delegate, root_span_name=_ROOT_SPAN)


# Module-level names for the framework-free steps in egma.otlp: the LiveKit
# tests monkeypatch them here, and _configure_provider reads them from here.
_setting = otlp.setting
_project_key = otlp.project_key
_trace_endpoint = otlp.trace_endpoint
_build_exporter = otlp.build_exporter
_flushed = otlp.flushed


def install(
    ctx: object,
    *,
    verb: str,
    endpoint: str | None,
    api_key: str | None,
    provider_reference: str = "",
) -> BatchSpanProcessor:
    """Export this worker's spans to Egma, once per process.

    ``provider_reference`` is the room a simulation runs in, or ``""`` for
    production. Returns the processor, so a caller with a second moment
    worth flushing at — a session closing — can ask for one.
    """
    trace_endpoint = _trace_endpoint(_setting(endpoint, "EGMA_URL", verb), verb)
    project_key = _project_key(_setting(api_key, "EGMA_API_KEY", verb), verb)
    add_shutdown_callback = getattr(ctx, "add_shutdown_callback", None)
    if not callable(add_shutdown_callback):
        raise ValueError(
            f"{verb} needs the LiveKit JobContext this job was given."
        )

    key_digest = hashlib.sha256(project_key.encode("utf-8")).digest()

    global _state
    with _state_lock:
        if _state is None:
            provider = _select_compatible_provider(verb, provider_reference)
            processor = _configure_provider(
                provider, trace_endpoint, project_key, verb, provider_reference
            )
            _state = _Export(
                verb=verb,
                endpoint=trace_endpoint,
                api_key_digest=key_digest,
                provider_reference=provider_reference,
                provider=provider,
                processor=processor,
            )
        elif (
            _state.endpoint != trace_endpoint
            or _state.api_key_digest != key_digest
        ):
            raise ValueError(
                f"{verb} is already configured with different settings in "
                "this process. Restart the worker after changing EGMA_URL or "
                "EGMA_API_KEY."
            )
        elif (
            _state.verb != verb
            or _state.provider_reference != provider_reference
        ):
            raise ValueError(
                f"{verb} cannot be configured in this process: it already "
                f"exports for {_state.verb}"
                + (
                    f" in room {_state.provider_reference!r}"
                    if _state.provider_reference
                    else " in a production room"
                )
                + ". A provider's resource is fixed when it is built, so one "
                "process exports one conversation. Run LiveKit with one job "
                "per process."
            )

        _register_shutdown_flush(ctx, _state.processor, verb)
        return _state.processor


def _livekit_provider() -> trace.TracerProvider:
    """Read the provider that LiveKit's dynamic tracer currently uses."""

    from livekit.agents.telemetry import tracer as livekit_tracer

    provider = getattr(livekit_tracer, "_tracer_provider", None)
    if provider is None:
        raise ValueError(
            "This LiveKit Agents version does not expose a compatible "
            "telemetry provider. Use a supported Egma SDK version."
        )
    return provider


def _select_compatible_provider(
    verb: str, provider_reference: str
) -> TracerProvider:
    """Reuse compatible telemetry and refuse to erase an existing provider.

    A customer's own provider is reused as it stands, resource and all.
    That resource cannot carry the room name — it was built before this
    SDK ran — which is exactly why the room name also rides every span
    through LiveKit's metadata seam below.
    """

    providers = {
        "LiveKit Agents": _livekit_provider(),
        "OpenTelemetry": trace.get_tracer_provider(),
    }
    concrete: dict[int, TracerProvider] = {}

    for owner, provider in providers.items():
        if isinstance(
            provider,
            (trace.ProxyTracerProvider, trace.NoOpTracerProvider),
        ):
            continue
        if not isinstance(provider, TracerProvider):
            raise ValueError(
                f"{verb} found an incompatible {owner} tracer provider. "
                "Configure one OpenTelemetry SDK TracerProvider before it."
            )
        concrete[id(provider)] = provider

    if len(concrete) > 1:
        raise ValueError(
            f"{verb} found different LiveKit Agents and OpenTelemetry tracer "
            "providers. Configure one shared provider before it."
        )
    if concrete:
        return next(iter(concrete.values()))

    attributes: dict[str, str] = {SERVICE_NAME: "livekit-agents"}
    if provider_reference:
        attributes[PROVIDER_REFERENCE] = provider_reference
    return TracerProvider(resource=Resource.create(attributes))


def _register_provider(provider: TracerProvider, provider_reference: str) -> None:
    """Make LiveKit and ordinary OpenTelemetry instrumentation share it.

    The room name goes through LiveKit's own ``metadata`` seam, which
    stamps it on every span the provider starts. That is the copy egma
    reads when the resource could not carry one.
    """

    global_provider = trace.get_tracer_provider()
    if isinstance(global_provider, trace.ProxyTracerProvider):
        trace.set_tracer_provider(provider)
    if provider_reference:
        set_tracer_provider(
            provider, metadata={PROVIDER_REFERENCE: provider_reference}
        )
        return
    set_tracer_provider(provider)


def _configure_provider(
    provider: TracerProvider,
    endpoint: str,
    api_key: str,
    verb: str,
    provider_reference: str,
) -> BatchSpanProcessor:
    """Attach one safe Egma processor and register its shared provider."""

    native_exporter = _build_exporter(endpoint, api_key, verb)
    exporter = (
        _SimulationEvidenceExporter(native_exporter)
        if provider_reference
        else native_exporter
    )
    try:
        processor = (
            BatchSpanProcessor(
                exporter, schedule_delay_millis=SIMULATION_BATCH_MILLIS
            )
            if provider_reference
            else BatchSpanProcessor(exporter)
        )
        _register_provider(provider, provider_reference)
        provider.add_span_processor(processor)
        return processor
    except Exception:
        # Telemetry libraries can include exporter state in exception text.
        # Close the unused exporter and replace that text with a safe message.
        try:
            exporter.shutdown()
        except Exception:
            pass
        raise ValueError(
            f"{verb} could not configure the Egma exporter. Check the "
            "worker's OpenTelemetry setup."
        ) from None


def _register_shutdown_flush(
    ctx: object, processor: BatchSpanProcessor, verb: str
) -> None:
    if getattr(ctx, _FLUSH_MARKER, False):
        return

    async def flush() -> None:
        await _flushed(processor, "job shutdown")

    ctx.add_shutdown_callback(flush)  # type: ignore[attr-defined]
    setattr(ctx, _FLUSH_MARKER, True)
